File: CSES/4. Graph Algorithm/Message Route.py
import sys, math, itertools, functools, collections
input = sys.stdin.readline
sys.setrecursionlimit(10**6)
rs  = lambda: input()
ri  = lambda: int(input())
rmi  = lambda: map(int, input().split())
ra = lambda: [int(x) for x in input().split()]
pa = lambda x: print (" ".join(map(str, x)))

n, m = rmi()
adj = [[] for _ in range(n+1)]
for _ in range(m):
	u, v = rmi()
	adj[u].append(v)
	adj[v].append(u)
 
step = 1
cur_level = [(1,[1])]
seen = set({1})
while cur_level:
	next_level = []
	for cur, path in cur_level:
		for nxt in adj[cur]:
			if nxt == n:
				print(step + 1)
				print(*(path[:] + [n]))
				exit()
			if nxt not in seen:
				seen.add(nxt)
				next_level.append((nxt, path + [nxt]))
	cur_level = next_level
	step += 1
print('IMPOSSIBLE')

File: CSES/4. Graph Algorithm/Message_Route.cpp
#include <bits/stdc++.h>
using namespace std;

#define ll long long
#define ld long double
#define fi first
#define se second
#define pb push_back
#define mp make_pair
#define sz(x) (int)(x).size()
#define rep(i,a,b) for(int i=(a);i<=(b);i++)
#define per(i,b,a) for(int i=(b);i>=(a);i--)
#define all(x) (x).begin(), (x).end()

template<typename T> bool chkMax(T&x, T y) { return (y>x)?x=y, 1:0; }
template<typename T> bool chkMin(T&x, T y) { return (y<x)?x=y, 1:0; }

template<typename T> void inline read(T&x){
    x=0;char c=getchar();bool flag=false;
    while(!isdigit(c)){if(c=='-')flag=true;c=getchar();}
    while(isdigit(c)){x=(x<<1)+(x<<3)+(c^48);c=getchar();}
    if(flag) x=-x;
}

const int MAXN = 1e6+5;
const int INF = 0x3f3f3f3f;

int n, m;
vector<int> adj[MAXN];

int main(){
    ios::sync_with_stdio(false); cin.tie(0); cout.tie(0);
    cin >> n >> m;
    rep(i,1,m){
        int u, v; cin >> u >> v;
        adj[u].pb(v);
        adj[v].pb(u);
    }
    int step = 1;
    vector<pair<int,vector<int>>> cur_level = {{1,{1}}};
    set<int> seen = {1};
    while(!cur_level.empty()){
        vector<pair<int,vector<int>>> next_level;
        for(auto [cur, path] : cur_level){
            for(auto nxt : adj[cur]){
                if(nxt == n){
                    cout << step+1 << "\n";
                    path.pb(n);
                    for(auto x : path)
                        cout << x << " ";
                    cout << "\n";
                    return 0;
                }
                if(!seen.count(nxt)){
                    seen.insert(nxt);
                    path.pb(nxt);
                    next_level.pb({nxt,path});
                    path.pop_back();
                }
            }
        }
        cur_level = next_level;
        step++;
    }
    cout << "IMPOSSIBLE\n";
    return 0;
}